src/subdomains/core/sell-crypto/process/buy-fiat-registration.service.ts

```diff
@@ -0,0 +1,109 @@
+    const newPayIns = await this.payInService.getNewPayIns();
+
+    if (newPayIns.length === 0) return;
+
+    const sellPayIns = await this.filterSellPayIns(newPayIns);
+
+    sellPayIns.length > 0 &&
+      console.log(
+        `Registering ${sellPayIns.length} new buy-fiat(s) from crypto pay-in(s) ID(s):`,
+        sellPayIns.map((s) => s[0].id),
+      );
+
+    await this.createBuyFiatsAndAckPayIns(sellPayIns);
+  }
+
+  //*** HELPER METHODS ***//
+
+  private async filterSellPayIns(allPayIns: CryptoInput[]): Promise<[CryptoInput, Sell][]> {
+    const routes = await this.sellRepository.find({
+      where: { deposit: Not(IsNull()) },
+      relations: ['deposit', 'user', 'user.userData'],
+    });
+
+    return this.pairRoutesWithPayIns(routes, allPayIns);
+  }
+
+  private pairRoutesWithPayIns(routes: Sell[], allPayIns: CryptoInput[]): [CryptoInput, Sell][] {
+    const result = [];
+
+    for (const payIn of allPayIns) {
+      const relevantRoute = routes.find(
+        (r) =>
+          payIn.address.address.toLowerCase() === r.deposit.address.toLowerCase() &&
+          payIn.address.blockchain === r.deposit.blockchain,
+      );
+
+      relevantRoute && result.push([payIn, relevantRoute]);
+    }
+
+    return result;
+  }
+
+  private async createBuyFiatsAndAckPayIns(payInsPairs: [CryptoInput, Sell][]): Promise<void> {
+    for (const [payIn, sellRoute] of payInsPairs) {
+      try {
+        const existingBuyFiat = await this.buyFiatRepo.findOne({ cryptoInput: { id: payIn.id } });
+
+        if (existingBuyFiat) {
+          const amlCheck = await this.payInService.acknowledgePayIn(payIn.id, PayInPurpose.SELL_CRYPTO, sellRoute);
+          existingBuyFiat.addAmlCheck(amlCheck);
+
+          await this.buyFiatRepo.save(existingBuyFiat);
+
+          continue;
+        }
+
+        // ignore DeFiChain AccountToUtxos for sell
+        if (payIn.txType === 'AccountToUtxos') {
+          console.log('Ignoring AccountToUtxos DeFiChain input on sell route. Pay-in:', payIn);
+          await this.payInService.ignorePayIn(payIn, PayInPurpose.SELL_CRYPTO, sellRoute);
+          continue;
+        }
+
+        await this.createNewBuyFiatAndAck(payIn, sellRoute);
+      } catch (e) {
+        console.error(`Error occurred during pay-in registration at buy-fiat. Pay-in ID: ${payIn.id}`, e);
+      }
+    }
+  }
+
+  private async createNewBuyFiatAndAck(payIn: CryptoInput, sellRoute: Sell): Promise<void> {
+    try {
+      const newBuyFiat = BuyFiat.createFromPayIn(payIn, sellRoute);
+
+      const amlCheck = await this.payInService.acknowledgePayIn(payIn.id, PayInPurpose.SELL_CRYPTO, sellRoute);
+
+      newBuyFiat.addAmlCheck(amlCheck);
+
+      await this.buyFiatRepo.save(newBuyFiat);
+    } catch (e) {
+      if (e instanceof SmallAmountException) {
+        await this.payInService.ignorePayIn(payIn, PayInPurpose.SELL_CRYPTO, sellRoute);
+
+        return;
+      }
+
+      throw e;
+    }
+  }
```

Comment:
same as above

Reply:
here unfortunately I have to keep saving after acknowledging, because we actually get amlCheck from acknowledging... but refactoring of two methods into one still makes sense